client/post-editor/editor-featured-image/dropzone.jsx

```diff
@@ -0,0 +1,114 @@
+/**
+ * External dependencies
+ */
+import React, { Component, PropTypes } from 'react';
+import { connect } from 'react-redux';
+import { head, uniqueId } from 'lodash';
+import { localize } from 'i18n-calypso';
+
+/**
+ * Internal dependencies
+ */
+import DropZone from 'components/drop-zone';
+import MediaActions from 'lib/media/actions';
+import MediaStore from 'lib/media/store';
+import MediaUtils from 'lib/media/utils';
+import PostActions from 'lib/posts/actions';
+
+import { receiveMedia, deleteMedia } from 'state/media/actions';
+import { editPost } from 'state/posts/actions';
+import { getSelectedSiteId } from 'state/ui/selectors';
+import { getEditorPostId } from 'state/ui/editor/selectors';
+
+class FeaturedImageDropZone extends Component {
+	static propTypes = {
+		site: PropTypes.object.isRequired,
+		post: PropTypes.object.isRequired,
+	};
+
+	static contextTypes = {
+		store: PropTypes.object
+	};
+
+	onFilesDrop = ( files ) => {
+		/**
+		 * Filter files for `image` media prefix and return the first image.
+		 *
+		 * At the moment we ignore all the other images that were dragged onto the DropZone
+		 */
+		const droppedImage = head( MediaUtils.filterItemsByMimePrefix( files, 'image' ) );
+
+		if ( ! droppedImage ) {
+			return false;
+		}
+
+		const transientMediaId = uniqueId( 'featured-image' );
+		const siteID = this.props.siteId;
+
+		const handleFeaturedImageUpload = () => {
+			const media = MediaStore.get( siteID, transientMediaId );
+			const isUploadInProgress = media && MediaUtils.isItemBeingUploaded( media );
+			const isFailedUpload = ! media;
+
+			if ( isFailedUpload ) {
+				this.props.deleteMedia( siteID, transientMediaId );
+			} else {
+				this.props.receiveMedia( siteID, media );
+			}
+
+			/**
+			 * File uploaded successfully, there is no longer need to listen for the changes.
+			 */
+			if ( media && ! isUploadInProgress ) {
+				MediaStore.off( 'change', handleFeaturedImageUpload );
+			}
+
+			/**
+			 * TODO: Redux way. What's left: figure out how to properly use `editPost`
+			 * and research if the whole FeaturedImage component has to be updated to
+			 * work properly with Redux.
+			 *
+			 * Right now `PostActions.edit` seems to be the best way to approach the problem.
+			 */
+			// this.props.editPost( siteID, this.props.postId, { featured_image: media.ID } );
+
+			// Cannot dispatch an action while in a dispatched action. Temporary(tm).
+			setTimeout( () => {
+				PostActions.edit( {
+					featured_image: media.ID
+				} );
+			}, 0 );
+		};
+
+		MediaStore.on( 'change', handleFeaturedImageUpload );
```

Comment:
It's used in several other places and I decided to follow that pattern, so I don't reinvent the wheel: [#](https://github.com/Automattic/wp-calypso/blob/470e739c4c41394cc0cca560067b06eb1d71512d/client/post-editor/editor-featured-image/preview-container.jsx#L35), [#](https://github.com/Automattic/wp-calypso/blob/7a19c4ea5c2ae605740870a7628d84a889d2e3f2/client/components/tinymce/plugins/media/plugin.jsx#L788), [#](https://github.com/Automattic/wp-calypso/blob/d9d0cd28a71b3aa3dc6dafda093e4fdd6f2b2ea6/client/my-sites/site-settings/site-icon-setting/index.jsx#L148-L148) ( the similarities on the last one are not coincidental :) )